amethyst_network/src/simulation/resource.rs

```diff
@@ -0,0 +1,210 @@
+    pub fn with_trusted_clients(mut self, clients: &[SocketAddr]) -> Self {
+        self.clients = clients.iter().map(|addr| Client::new(*addr)).collect();
+        self
+    }
+
+    /// Returns a slice of the tracked clients
+    pub fn clients(&self) -> &[Client] {
+        &self.clients
+    }
+
+    /// Set the server address
+    pub fn set_server_addr(&mut self, server_addr: Option<SocketAddr>) {
+        self.server_addr = server_addr;
+    }
+
+    /// Return a mutable reference to the socket if there is one configured.
+    pub fn get_socket_mut(&mut self) -> Option<&mut S> {
+        self.socket.as_mut()
+    }
+
+    /// Set the bound socket to the `NetworkSimulationResource`
+    pub fn set_socket(&mut self, socket: S) {
+        self.socket = Some(socket);
+    }
+
+    /// Drops the socket from the `NetworkSimulationResource`
+    pub fn drop_socket(&mut self) {
+        self.socket = None;
+    }
+
+    /// Returns whether or not the `NetworkSimulationResource` has a bound socket
+    pub fn has_socket(&self) -> bool {
+        self.socket.is_some()
+    }
+
+    /// Returns the server address if one was set
+    pub fn server_addr(&self) -> Option<SocketAddr> {
+        self.server_addr
+    }
+
+    /// Returns whether or not this `NetworkSimulationResource` was created as a 'server'
+    pub fn is_server(&self) -> bool {
+        self.is_server
+    }
+
+    /// Returns whether or not this `NetworkSimulationResource` was created as a 'client'
+    pub fn is_client(&self) -> bool {
+        !self.is_server
+    }
+
+    /// Create a `Message` with the default guarantees provided by the `Socket` implementation and
+    /// pushes it onto the messages queue to be sent on next sim tick.
+    pub fn send(&mut self, payload: &[u8]) {
+        self.send_with_requirements(
+            payload,
+            S::default_requirement(),
+            UrgencyRequirement::OnTick,
+        );
+    }
+
+    /// Create a `Message` with the default guarantees provided by the `Socket` implementation and
+    /// pushes it onto the messages queue to be sent immediately.
+    pub fn send_immediate(&mut self, payload: &[u8]) {
+        self.send_with_requirements(
+            payload,
+            S::default_requirement(),
+            UrgencyRequirement::Immediate,
+        );
+    }
+
+    /// Create and queue a `Message` with the specified guarantee
+    pub fn send_with_requirements(
+        &mut self,
+        payload: &[u8],
+        delivery: DeliveryRequirement,
+        timing: UrgencyRequirement,
+    ) {
+        let message = Message::new(payload, delivery, timing);
+        self.messages.push_back(message);
+    }
+
+    /// Returns true if there are messages enqueued to be sent.
+    pub fn has_messages(&self) -> bool {
+        !self.messages.is_empty()
+    }
+
+    /// Drains the messages queue and returns the drained messages
+    pub fn drain_messages(&mut self, mut filter: impl FnMut(&mut Message) -> bool) -> Vec<Message> {
```

Comment:
We drain messages based on the given priority. What do you think of the idea to use a [priority queue](https://crates.io/crates/priority-queue) instead?

Reply:
Would prefer using the std lib rather than bringing in a dependency for a single data structure. That is, unless we profile and find that it is not performing adequately.

Reply:
We might as well consider writing our own, by just copying some of their code. Shouldn't be that many lines.